### PR TITLE
CON-1094_ DC_API_integration_learner_data

### PR DESCRIPTION
**Pull request by ChrisJWoodcock**
The academic year DC API endpoint is now returning both 1920 and 2021 academic years, however there will be no data exposed from the 2021 endpoint until it is made live which is in August of each year.

### CODE REVIEW

**Comment by ChrisJWoodcock** [comment]
This has dependencies for CON-1154 and CON-1217 for testing.

File: src/SFA.DAS.Assessor.Functions/Domain/EpaoDataSync/EpaoDataSyncLearnerService.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                    {
+                        ImportLearnerDetails = filteredLearners
+                    };
+
+                    var response = await _assessorServiceApiClient.ImportLearnerDetails(request);
+                    response?.LearnerDetailResults?.ForEach(ld =>
+                    {
+                        ld?.Errors?.ForEach(e =>
+                        {
+                            _logger.LogDebug($"Unable to export Learner Details due to '{ld.Outcome}' '{e}'");
```
Comment (cofaulco) [inline src/SFA.DAS.Assessor.Functions/Domain/EpaoDataSync/EpaoDataSyncLearnerService.cs]: we call **Import**LearnerDetails however this log line says unable to **Export**. Should these be consistent?

Comment (ChrisJWoodcock) [reply src/SFA.DAS.Assessor.Functions/Domain/EpaoDataSync/EpaoDataSyncLearnerService.cs]: I wondered about that when I wrote it I suppose 'request to import failed' would better...

File: src/SFA.DAS.Assessor.Functions.ExternalApis/DataCollection/DataCollectionServiceApiClient.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        {
+            var requestUri = learnersRequestUri
+                + (aimType != null
+                    ? $"&aimType={aimType.Value}"
+                    : string.Empty)
+                + (standardCode != null
+                    ? $"&standardCode={standardCode.Value}"
+                    : string.Empty)
+                + (fundModels != null
+                    ? string.Join(string.Empty, fundModels.ConvertAll(p => $"&fundModel={p}").ToArray())
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.Assessor.Functions.ExternalApis/DataCollection/DataCollectionServiceApiClient.cs]: Ok

File: src/SFA.DAS.Assessor.Functions.ExternalApis/ApiClientBase.cs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        {
+            Client = httpClient;
+
+            Client.DefaultRequestHeaders.Accept.Clear();
+            Client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", tokenService.GetToken());
+            Client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+
+            _logger = logger;
+
+            _retryPolicy = HttpPolicyExtensions
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.Assessor.Functions.ExternalApis/ApiClientBase.cs]: Ok

File: src/SFA.DAS.Assessor.Functions.ExternalApis/ApiClientBase.cs [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+using Polly.Extensions.Http;
+using Polly.Retry;
+using SFA.DAS.Assessor.Functions.ExternalApis.Exceptions;
+using SFA.DAS.Assessor.Functions.ExternalApis.Interfaces;
+
+namespace SFA.DAS.Assessor.Functions.ExternalApis
+{
+    public abstract class ApiClientBase : IDisposable
+    {
+        protected HttpClient Client;
```
[the remark this answers is not in the document]
Comment (ChrisJWoodcock) [reply src/SFA.DAS.Assessor.Functions.ExternalApis/ApiClientBase.cs]: Ok

**Comment by ChrisJWoodcock** [comment]
> Rather than creating a class to handle the queuing of messages, you can use output bindings on the timer function to queue the messages.
> 
> See the section on output: https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus

Those examples are basic calculation functions; the provider messages are quite complicated and it is easier to test by including the functionality in a service class which is dependant on a message queuing class which I mock for the tests.

**Comment by cofaulco** [comment]
> > Rather than creating a class to handle the queuing of messages, you can use output bindings on the timer function to queue the messages.
> > See the section on output: https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus
> 
> Those examples are basic calculation functions; the provider messages are quite complicated and it is easier to test by including the functionality in a service class which is dependant on a message queuing class which I mock for the tests.

The service could still build the list of messages and return them. The top level function could just be responsible for queuing them. This would keep the testability of the service

**Comment by ChrisJWoodcock** [comment]
> What was the reason for using Environment.GetVariable() rather than IConfiguration ??
> 
> Also the EpaoServicebusConnectionString can now be removed from the readme localsettings.json setup.

This use of Environment.GetVariable() is the recommended approach from MS on getting both configuration and environment variables - it also allows the IConfiguration to be removed from startup and being passed around to other classes.

I'll edit the Readme  - I thought I did already as I meant to do it...

File: src/SFA.DAS.Assessor.Functions.ExternalApis/DataCollection/Authentication/DataCollectionTokenService.cs [hunk 5]
```diff
@@ -1,29 +1,42 @@
...
+                    .WithAuthority(new Uri(_dataCollectionApiAuthenticationOptions?.Value.Authority))
+                    .Build();
+
+            string[] scopes = new string[] { scope };
 
-            var t = _dataCollectionServiceAnonymousApiClient.GetToken(tokenRequest);
+            AuthenticationResult result;
+            try
+            {
+                result = app.AcquireTokenForClient(scopes).ExecuteAsync().Result;
```
Comment (cofaulco) [inline src/SFA.DAS.Assessor.Functions.ExternalApis/DataCollection/Authentication/DataCollectionTokenService.cs:32]: Should we await this async call and make this method async, returning a Task ?

**Comment by ChrisJWoodcock** [comment]
> How come we don't use the Shared nuget sfa.das.http to build the HttpClient with AAD as with other services?

I didn't know the shared package existed; however on investigation the shared package uses a version 1.0 endpoint for Azure AD Authentication Library for .NET (ADAL.NET) and the DC API team has supplied a Scope because they are using the version 2.0 endpoint with Microsoft Authentication Library for .NET (MSAL.NET); so the version as written is correct - when EFSA system migrate to use Scopes the shared package could be updated and used.

File: src/SFA.DAS.Assessor.Functions.ExternalApis/DataCollection/Authentication/DataCollectionTokenService.cs (hunk 5 above)
Comment (ChrisJWoodcock) [reply src/SFA.DAS.Assessor.Functions.ExternalApis/DataCollection/Authentication/DataCollectionTokenService.cs]: This has now been Async using a DelegatingHandler from the HttpClient pattern

**Pull request merged**